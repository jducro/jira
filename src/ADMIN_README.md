## Installation

1. Pre-requisites

    You will need to generate an openssl certificate to authenticate against your Jira server from the app. You skip this step if your system has openssl installed, otherwise keep on reading for instructions on how to install it
    
    1. Debian\Ubuntu Linux:
    
        ```
            sudo apt-get install openssl
        ```
    
    2. CentOS \ Redhat
    
        ```
            sudo yum -y install openssl
        ```         
    
    3. macOS
    
        ```
         brew install openssl
        ```
    
    4. Windows 
    
        There are many alternatives which provide the binaries required for Windows (see the  [official wiki](https://wiki.openssl.org/index.php/Binaries) ), but we recommend going to [https://slproweb.com/products/Win32OpenSSL.html](https://slproweb.com/products/Win32OpenSSL.html) and downloading the installer named `Win64 OpenSSL v1.0.2o` or the latest version if available. Run the installer and follow the instructions.
        
2. Generate public and private keys for secure communications with your JIRA server
    
    These two keys are required for authentication with your JIRA server. The private key will be stored by your Deskpro instance, while the public key remains on the Jira server.
    
    Open a terminal or a command prompt if on windows and type the following commands:
      
    1. Generate the private key
        
        ```
        openssl genrsa -out private.pem 2048
        ```
        
    2. Extract the public key
        
        ```
        openssl rsa -in private.pem -outform PEM -pubout -out public.pem
        ```    
3. Configure the app connection in JIRA

    1. Navigate to `Applications` page
     
        Log in to your JIRA instance, click the settings menu (small cog wheel next to your profile bage on the top right corner of the page) and choose `Applications`. You should arrive at a page that looks like this:
        
        ![Applications page](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/001-application-links.small.png)
        
    2. Configure an Application Link
    
        From the `Integrations` section choose `Application links` and fill the url of your Deskpro instance in the `Create new link` input box:
        
        ![Configure application links](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/002-configure-application-link.small.png))
        
        If you receive a warning the url being invalid although you enter it correctly, it's safe to click `Continue` and ignore it
        
        ![Ignore invalid url warning](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/003-skip-invalid-url.small.png))
    
    3. Link Applications
        
        Fill in the fields as shown in the image below and make sure you enable `Create incoming link` then click `Continue`
        
        ![Link applications](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/004-incoming-link.small.png))
        
    4. Fill in the incoming link details
        
        You'll have to **remember** the `Consumer Key` because you'll need it when installing the app in Deskpro. 
        Copy the contents of the `public.pem` file we generated at the beginning in the `Public Key` field.

4. Install the JIRA App in Deskpro

    1. Install the application
     
        Login to your Deskpro instance, then navigate to `Admin`, then click `Apps` from the `Apps` menu. Scroll down and choose `JIRA v2` then click `Install App`
        
        ![Install application](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/006-install-app.small.png))
        
    2. Configure the application
    
        - Fill the `Rsa private key` field with the contents of the `private.pem` file we generated in step 2
        - Fill the `Rsa public key` field with the contents of the `public.pem` file we generated in step 2        
        - Fill the `JIRA instance url` field with the url to the JIRA server we used in step 3, for example: `https://your-company.atlassian.net` if you are using the cloud version        
        - Fill the `JIRA Consumer Key` field with the value of the `Consumer Key` field you remember from step 3
        
        ![Install application](https://raw.githubusercontent.com/DeskproApps/jira/master/docs/install-guide/007-configure-app.small.png)
        
        Click `Update settings` and give your consent in the pop-up that appears to authorize the connection between the app and JIRA. 
        Congratulations! You can now link JIRA issues from inside Deskpro.          
    
 
