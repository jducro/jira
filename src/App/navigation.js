
const ROUTE_HOME  =     'linkedIssues';
const ROUTE_BROWSE =    'browseIssues';
const ROUTE_CREATE =    'createIssue';
const ROUTE_EDIT =      'editIssue';
const ROUTE_SIGNIN  =   'signIn';
const ROUTE_LOADING =  'loading';

export const routesEnum =
{
  ROUTE_BROWSE,
  ROUTE_CREATE,
  ROUTE_EDIT,
  ROUTE_HOME,
  ROUTE_SIGNIN,
  ROUTE_LOADING
};

export function gotoHome(args)
{
  return function(history) {
    history.push('linkedIssues', args);
    history.go(1);
  };
}

export function gotoCreate(args)
{
  return function(history) {
    history.push('createIssue', args);
    history.go(1);
  }
}

export function gotoBrowse(args)
{
  return function(history) {
    history.push('browseIssues', args);
    history.go(1);
  }
}

export function gotoEdit(args)
{
  return function(history) {
    history.push('editIssue', args);
    history.go(1);
  }
}

export function gotoSignIn(args)
{
  return function(history) {
    history.push('signIn', args);
    history.go(1);
  }
}
