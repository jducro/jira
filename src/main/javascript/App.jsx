import React from 'react';
import ReactDOM from 'react-dom';

export default class App extends React.Component
{
  static propTypes = { dpapp: React.PropTypes.object.isRequired };

  shouldComponentUpdate() { return false; }

  render() { return (<div>Hello world</div>); }
}
