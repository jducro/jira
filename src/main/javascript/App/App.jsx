import React from 'react';
import PropTypes from 'prop-types';
import { CreateMetadataFinder } from '../Jira'
import { createThrottle, createReducerChain } from '../Infrastructure'
import { UI } from '../UI';
import { JiraService } from '../Jira';
import { Routes } from './Routes';


const reduce = createReducerChain([
  require('./Services').reducer,
  require('../BrowseIssues').reducer,
  require('../LinkIssues').reducer,
  require('../CreateIssue').reducer
]);

export class App extends React.Component
{
  static propTypes = {

    dpapp: PropTypes.object.isRequired,

    ui: PropTypes.object.isRequired,

    /**
     * Instance of sdk route.
     * @see https://deskpro.gitbooks.io/deskpro-apps/content/api/props/route.html
     */
    route:   PropTypes.object
  };

  static childContextTypes = {

    loadJiraCreateMeta: PropTypes.func,

    loadJiraEditMeta: PropTypes.func,

    ticket: PropTypes.func,

    dispatch: PropTypes.func,
  };

  constructor(props)
  {
    super(props);
    this.init();
  }

  init()
  {
    this.state = {

      appReady: false,

      jiraInstanceUrl: '',

      jiraCards: [],

      linkedIssues: [],

      foundIssues: [],

      handleReply: {
        reply: false,
        strategy: 'none'
      }
    };

    const { dpapp, tabData } = this.props ;

    this.childContext = {

      dispatch: this.dispatch,

      loadJiraCreateMeta: this.loadJiraCreateMeta.bind(this),

      loadJiraEditMeta: this.loadJiraEditMeta.bind(this),

      ticket:  () => ({
        url: dpapp.context.tabUrl,
        id: dpapp.context.entityId,
        title: tabData.original_subject
      })
    };
  }

  componentDidMount()
  {
    // load state, initialize services
    const { ui } = this.props ;
    this.loadState()
      .then(state => {
        this.setState(state);
        return state;
      })
      .then(this.synchronizeState.bind(this))
      .then(state => {
        this.subscribeToHelpdeskEvents();
        this.renderHelpdeskUI();
        return state;
      })
      .then(state => {
        const jiraService = this.createJiraService();
        return jiraService.verifyAccess()
          .then(() => {
            this.props.route.to(Routes.linkedIssues);
            return state;
          })
          .catch(err => {
            this.props.route.to(Routes.signIn);
            return state;
          })
        ;
      })
      .then((state) => {
        this.setState({ ...state, appReady: true });
      })
      .catch(ui.error)
    ;
  }

  componentDidUpdate(prevProps, prevState)
  {
    const { context } = this.props.dpapp;

    if (prevState.linkedIssues !== this.state.linkedIssues) {
      const allJiraCards = [].concat(this.state.jiraCards).concat(this.state.linkedIssues.map(issue => issue.key));
      const uniqueCards = allJiraCards.filter(function(elem, pos, arr) {
        return arr.indexOf(elem) == pos;
      });
      context.customFields.setAppField('jiraCards', uniqueCards);
    }
  }

  dispatch = action =>
  {
    if (typeof action === 'function') {
      const jiraService = this.createJiraService();
      return action({ dispatch: this.dispatch, jiraService, dpapp: this.props.dpapp });
    }

    if (typeof action === 'object') {
      const state = reduce(this.state, action);
      if (state !== this.state) {
        this.setState(state);
      }
    }
  };

  synchronizeState(state)
  {
    // synchronize jiraCards
    const { jiraCards } = state;
    if (0 === jiraCards.length) {
      return jiraCards;
    }

    const jiraService = this.createJiraService( state.jiraInstanceUrl );
    const { context } = this.props.dpapp;

    return jiraService.readAllIssues(jiraCards).then(issues => {
      return {...state, linkedIssues: issues};
    });
  }

  subscribeToHelpdeskEvents()
  {
    this.props.dpapp.subscribe('context.ticket.update-success', createThrottle(this.renderHelpdeskUI.bind(this), 500));
    this.props.dpapp.subscribe('context.ticket.reply', this.onTicketReply.bind(this));
    this.props.dpapp.subscribe('context.ticket.reply-success', this.onTicketReplySuccess.bind(this));
  }

  renderHelpdeskUI()
  {
    const { deskproWindow } = this.props.dpapp ;
    const { entityId: ticketId } = this.props.dpapp.context;
    const domRootId = `app-jira-${ticketId}`;

    const insertQuery = {
      parent: `#ticket${ticketId}-reply-controls`,
      markup: `<div class="cell" id="${domRootId}">
                  <div class="inner-cell" style="padding-left: 11px;">
                    <input type="checkbox" id="ticket${ticketId}-reply-with-jira" />
                  </div>
                  <div class="inner-cell" style="position: relative;">
                      <select id="ticket${ticketId}-reply-with-jira-strategy" style="min-width: 170px;" data-style-type="icons">
                      <option value="add-comment" selected="selected">Send comment to JIRA</option>
                      <option value="new-issue">Send to JIRA as new issue</option>
                    </select>
                  </div>
              </div>`
    };

    deskproWindow.domQuery({ type: 'exists', selector: `#${domRootId}`})
      .then(value => {
        if (false === value.exists) {
          deskproWindow.domInsert(insertQuery);
        }
    });
  }

  onTicketReply(response, data)
  {
    const { deskproWindow } = this.props.dpapp ;
    const { entityId: ticketId } = this.props.dpapp.context;

    const queries = [{
      name: 'reply',
      type: 'valueOf',
      selector: `#ticket${ticketId}-reply-with-jira`
    },
      {
        name: 'strategy',
        type: 'valueOf',
        selector: `#ticket${ticketId}-reply-with-jira-strategy`
      },
    ];

    const reduceQueryResults = (o, {name, value}) => {
      o[name] = value;
      return o;
    };

    deskproWindow.domQuery(queries)
      .then(results => {
        const handleReply = results.reduce(reduceQueryResults, { reply: false });
        const onAfterSetState = () => { response({ allowReply: true }); };
        this.setState({ handleReply }, onAfterSetState);
    });
  }

  onTicketReplySuccess(data)
  {
    const { message } = data;
    const { reply, strategy } = this.state.handleReply;

    if (reply && strategy === 'add-comment') {
      const { ui } = this.props;
      return this.addComment(message).catch(ui.error);
    }

    if (reply && strategy === 'new-issue') {
      const { route } = this.props;
      route.to(Routes.createIssue, { comment: message });
    }
  }

  loadState()
  {
    const { dpapp } = this.props ;
    const state = {
      jiraInstanceUrl: '',
      jiraCards: []
    };

    return dpapp.storage.getAppStorage('jiraInstanceUrl')
      .then(value => {
        state.jiraInstanceUrl = value;
        return dpapp.context.customFields.getAppField('jiraCards', [])
      })
      .then(cards => {
        state.jiraCards = cards
      })
      .then(() => {
        return state;
      }).catch(e => {
        console.error(e);
      })
    ;
  }

  /**
   * @return {JiraService}
   */
  createJiraService( jiraInstanceUrl )
  {
    const { dpapp } = this.props ;
    const instanceUrl = jiraInstanceUrl || this.state.jiraInstanceUrl;

    return new JiraService({
      httpClient: dpapp.restApi.fetchCORS.bind(dpapp.restApi),
      instanceUrl
    });
  }

  /**
   * @return {Promise.<CreateMetadataFinder>}
   */
  loadJiraCreateMeta()
  {
    const meta = window.sessionStorage.getItem('createMeta');
    if (meta) {
      return Promise.resolve(
        new CreateMetadataFinder(JSON.parse(meta))
      );
    }

    return this.createJiraService().loadCreateMeta().then(meta => {
      window.sessionStorage.setItem('createMeta', JSON.stringify(meta));
      return new CreateMetadataFinder(meta);
    });
  }

  /**
   * @return {Promise.<{}>}
   */
  loadJiraEditMeta(issue)
  {
    return this.createJiraService().loadEditMeta(issue).then(meta => {
      return meta;
    });
  }

  /**
   * @param {string} comment
   * @return {*}
   */
  addComment(comment)
  {
    const { linkedIssues } = this.state;
    if (linkedIssues.length === 0) {
      return Promise.reject(new Error('there are no linked issues'));
    }
    const jiraService = this.createJiraService();

    const promises = linkedIssues.map(issue => jiraService.createComment(issue, comment).then(
      result => ({ status:'success', result }),
      err => ({ status:'success', result: err })
    ));

    return Promise.all(promises).then(results => results.filter(x => x.status === 'success').map(x => x.result));
  }

  getChildContext() { return this.childContext; }

  render()
  {
    if (this.state.appReady === false) {
      return null;
    }

    const { linkedIssues, foundIssues } = this.state;
    const { route } = this.props;

    return (<UI route={route} dispatch={this.dispatch} linkedIssues={linkedIssues} foundIssues={foundIssues}/>);
  }
}
