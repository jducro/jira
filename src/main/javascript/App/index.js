export {Routes} from './Routes';
export {App} from './App';
