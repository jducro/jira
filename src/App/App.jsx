import React from 'react';
import PropTypes from 'prop-types';
import { Router, Route, Switch,  } from 'react-router'
import { Loader, Panel, Level, Action, ActionBar } from '@deskpro/apps-components';

import { ScreenCreateIssue, ScreenEditIssue } from '../CreateIssue';
import { ScreenBrowseIssues } from '../BrowseIssues';
import { ScreenLinkIssues } from '../LinkIssues';
import { ScreenAuth } from '../Security';

import { routesEnum, gotoHome, gotoSignIn } from './navigation';
import { DeskproPortal } from '../Comments';
import { verifyAccess } from '../Security'
import { reduxConnector } from "../infrastructure";
import {gotoBrowse, gotoCreate} from "./index";

export class App extends React.PureComponent
{
  static propTypes = {
    dpapp:        PropTypes.object.isRequired,
    history:      PropTypes.object,
    navigator:    PropTypes.func.isRequired,
    verifyAccess: PropTypes.func.isRequired,
  };

  componentDidMount()
  {
    const { verifyAccess, dpapp, navigator } = this.props;

    verifyAccess()
      .then(() => navigator()(gotoHome))
      .catch(err => { navigator()(gotoSignIn); })
      .catch(dpapp.ui.showErrorNotification)
    ;
  }

  gotoHome = () => {
    const { navigator } = this.props;
    navigator()(gotoHome)
  };

  gotoCreate = () => {
    const { navigator } = this.props;
    navigator()(gotoCreate)
  };

  gotoBrowse = () => {
    const { navigator } = this.props;
    navigator()(gotoBrowse)
  };

  render()
  {
    const { navigator, dpapp } = this.props;
    return [
      <DeskproPortal navigator={navigator} dpapp={dpapp}/>,
      this.renderRouter()
    ];
  }

  renderRouter()
  {
    const { history } = this.props;

    return <Router history={history} >
      <Switch>
        <Route path={routesEnum.ROUTE_SIGNIN} render={this.renderSignIn} />
        <Route path={routesEnum.ROUTE_CREATE} render={this.renderScreenCreateIssue} />
        <Route path={routesEnum.ROUTE_HOME} render={this.renderScreenLinkIssues} />
        <Route path={routesEnum.ROUTE_BROWSE} render={this.renderScreenBrowseIssues} />
        <Route path={routesEnum.ROUTE_EDIT} render={this.renderScreenEdit} />
        <Route path={routesEnum.ROUTE_LOADING} render={this.renderLoader} />
        <Route render={this.renderLoader} />
      </Switch>
    </Router>
  }

  renderLoader = () =>
  {
    return <Level><Loader /></Level>;
  };

  renderSignIn = ({match, location, history}) =>
  {
    const { navigator, dpapp } = this.props;
    return <ScreenAuth dpapp={dpapp} navigator={navigator} />
  };


  renderScreenCreateIssue = ({match, location, history}) =>
  {
    let comment = null;
    if (location.state) {
      comment = location.state.comment
    }

    const { navigator, dpapp } = this.props;
    return <Panel title={"Create an issue"} border={"none"}>
      <Action icon={'close-heavy'} label={"Go back"} onClick={this.gotoHome}/>
      <ScreenCreateIssue dpapp={dpapp} navigator={navigator} comment={comment} />
    </Panel>;
  };

  renderScreenLinkIssues = ({match, location, history}) =>
  {
    const { navigator, dpapp } = this.props;

    return [
      <ActionBar key={"linked-issues-actions"} title={"Linked Issues"}>
        <Action icon={'add'} label={"Add"} onClick={this.gotoCreate}/>
        <Action icon={'search'} label={"Find"} onClick={this.gotoBrowse}/>
      </ActionBar>,
      <ScreenLinkIssues key={"screen-link-issues"} dpapp={dpapp} navigator={navigator} onCancelRoute={gotoHome} />
    ]
  };

  renderScreenBrowseIssues = ({match, location, history}) =>
  {
    const { navigator, dpapp } = this.props;

    return <Panel title={"Find issues"} border={"none"}>
      <Action icon={'close-heavy'} label={"Go back"} onClick={this.gotoHome}/>
      <ScreenBrowseIssues dpapp={dpapp} navigator={navigator} />
    </Panel>;

  };

  renderScreenEdit = ({match, location, history}) =>
  {
    const { issue, onCancelRoute } = location.state;
    const { navigator, dpapp } = this.props;

    return (
      <Panel title={"Edit issue"} border={"none"}>
        <Action icon={'close-heavy'} label={"Go back"} onClick={this.gotoHome}/>
        <ScreenEditIssue dpapp={dpapp} navigator={navigator} issue={issue} />
      </Panel>
    );
  }

}

export default reduxConnector(App, { verifyAccess }, {});
