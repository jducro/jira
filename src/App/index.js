import { default as App } from './App'

export { App };
export { gotoBrowse, gotoCreate, gotoEdit, gotoHome, gotoSignIn, routesEnum } from './navigation';
