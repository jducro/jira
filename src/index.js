import { AppFrame } from '@deskpro/apps-components';
import { createApp } from '@deskpro/apps-sdk';
import ReactDOM from 'react-dom';
import React from 'react';
import { Provider } from "react-redux";
import { createMemoryHistory as createHistory } from "history";

import './styles.css';
import { App } from './App';
import { createNavigator } from './infrastructure';
import configureStore from './store';

const history = createHistory({
  initialEntries: ["loading"],
  initialIndex: 0
});

createApp(dpapp => props => {
  configureStore(dpapp).then(store =>
    ReactDOM.render(
      <AppFrame {...props}>
        <Provider store={store}>
          <App dpapp={dpapp} history={history} navigator={createNavigator(history)} />
        </Provider>
      </AppFrame>,
      document.getElementById('root')
    )
  );
});
