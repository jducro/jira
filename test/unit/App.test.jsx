import React from 'react';
import renderer from 'react-test-renderer';
import { default as bootstrap } from './helpers'
import { Provider } from "react-redux";
import { App, routesEnum } from '../../src/App';

test('successfully render the loading screen', done => {

  const httpClient = () => Promise.resolve(null);
  const { store, dpapp, history, navigator } = bootstrap({ httpClient, initialRoute: routesEnum.ROUTE_LOADING });

  const component = renderer.create(
    <Provider store={store}>
      <App dpapp={dpapp} history={history} navigator={navigator}/>
    </Provider>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  done();
});

test('successfully render the home screen', done => {

  const httpClient = () => Promise.resolve(null);
  const { store, dpapp, history, navigator } = bootstrap({ httpClient, initialRoute: routesEnum.ROUTE_HOME });

  const component = renderer.create(
    <Provider store={store}>
      <App dpapp={dpapp} history={history} navigator={navigator}/>
    </Provider>
  );

  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
  done();
});
