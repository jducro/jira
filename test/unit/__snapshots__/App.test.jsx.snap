// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`successfully render the home screen 1`] = `
Array [
  <div
    className="dp-ActionBar dp-Level"
>
    <div
        className="dp-LevelLeft"
    >
        <span
            className="dp-ActionBarTitle"
        >
            Linked Issues
        </span>
    </div>
    <div
        className="dp-ActionList dp-LevelRight"
    >
        <a
            className="dp-ActionListItem dp-ActionListIcon dp---is-hoverable"
            onClick={[Function]}
            title=""
        >
            <i
                className="dp-IconPlus"
            />
            <span
                className="dp-ActionListLabel"
            >
                Add
            </span>
        </a>
        <a
            className="dp-ActionListItem dp-ActionListIcon dp---is-hoverable"
            onClick={[Function]}
            title=""
        >
            <i
                className="dp-IconSearch"
            />
            <span
                className="dp-ActionListLabel"
            >
                Find
            </span>
        </a>
    </div>
</div>,
  <p>
    You haven't linked any issues to this ticket. 
</p>,
]
`;

exports[`successfully render the loading screen 1`] = `
<div
  className="dp-Level"
>
  <div
    className="dp-Loader dp-Loader--isActive"
  />
</div>
`;
